### PR TITLE
Add support for `EXT_mesh_primitive_edge_visibility`

### LINKED ISSUE #1809: Investigate support for `EXT_mesh_primitive_edge_visibility`
### Feature

`EXT_mesh_primitive_edge_visibility` is a proposed extension that encodes information to allow engines to render visible edges around 3D objects.

Image

This issue is for investigating (and hopefully implementing) support for this extension in Unreal Engine. The edges require a separate drawing pass in CesiumJS; perhaps the Unreal equivalent is to do a post-process ?

- [Spec](https://github.com/CesiumGS/glTF/tree/pmc/edge-visibility/extensions/2.0/Vendor/EXT_mesh_primitive_edge_visibility)
- [CesiumJS Implementation](https://github.com/CesiumGS/cesium/pull/12859/)

### PR DESCRIPTION
**Pull request by j9liu**
## Description

~Depends on #1836, so merge that first.~

This PR supports rendering edges from the [`EXT_mesh_primitive_edge_visibility`](https://github.com/CesiumGS/glTF/tree/pmc/edge-visibility/extensions/2.0/Vendor/EXT_mesh_primitive_edge_visibility) glTF extension, which is often used to enhance the appearance of BIM/CAD models, shown below:

| Without Edges | With |
| - | - |
|HighresScreenshot00000 | HighresScreenshot00003 |
| HighresScreenshot00001 | HighresScreenshot00004 |
| HighresScreenshot00002 | HighresScreenshot00005 |

## Issue number or link

Closes #1809.

## Author checklist

- [X] I have submitted a [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) (only needed once).
- [x] I have done a full self-review of my code.
- [X] I have updated `CHANGES.md` with a short summary of my change (for user-facing changes).
- ~[ ] I have added or updated unit tests to ensure consistent code coverage as necessary.~
- ~[ ] I have updated the documentation as necessary.~

## Remaining Tasks

- [x] Parse the edge material to color edges correctly.
- [X] Support `lineString` + material overrides - #1842 
- ~[ ] Investigate if multi-pass approach mitigates z-fighting artifacts:~
image

Aside from these cases where a bunch of tiny edges are bunched together, they look pretty good as is, so we'll hold off on the multi-pass approach for now.

## Testing plan

Reach out to me for the pictured test dataset! It should be as simple as loading any other tileset into Unreal, just confirm that the edges show up.

## Reviewer checklist

Thank you for taking the time to review this PR. By approving a PR you are taking as much responsibility for these changes as the author.

As you review, please go through the checklist below:

- [ ] Review and run all parts of the test plan on this branch and verify it matches expectations.
    - [ ] If the issue is a bug please make sure you can reproduce the bug in the main branch and then checkout this branch to make sure it actually solved the issue.
- [ ] Review the code and make sure you do not have any remaining questions or concerns. You should understand the code change and the chosen approach. If you are not confident or have doubts about the code, please do not hesitate to ask questions.
    - [ ] Code should follow the [C++ Style Guide](https://cesium.com/learn/cesium-native/ref-doc/style-guide.html)
- [ ] Review the unit tests and make sure there are no missing tests or edge cases.
- [ ] Review documentation changes and updates to `CHANGES.md` to make sure they accurately cover the work in this PR.
- [ ] Verify that the [Contributor License Agreement](https://github.com/CesiumGS/community/tree/main/CLAs) has been submitted, if needed.

### CODE REVIEW

**Comment by j9liu** [comment]
I added a function to parse the base color of the material used by the `EXT_mesh_primitive_edge_visibility` extension, or to otherwise fallback on the primitive's own material.

However, the test data I'm using doesn't contain such a material, so it defaults to the default glTF material that is all white. Not as cool looking, but technically correct behavior:

image

In any case, I think this is in a state to do an early review while I work on the other two Remaining Task items as follow-up PRs.

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+  FStaticMeshVertexBuffer& edgeVertexBuffer =
+      lodResources.VertexBuffers.StaticMeshVertexBuffer;
+  edgeVertexBuffer.Init(totalEdgeVertices, numTexCoords, false);
+
+  TArray<uint32> indices;
+  indices.Reserve(totalEdgeVertices);
+
+  glm::vec3 minPosition{std::numeric_limits<float>::max()};
+  glm::vec3 maxPosition{std::numeric_limits<float>::lowest()};
```
Comment (j9liu) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: No longer needed, should be removed.

File: Shaders/Private/CesiumVoxelTemplate.usf [hunk 2]
```diff
@@ -74,7 +74,7 @@ struct VoxelMegatextures
     {
       // Since glTFs are y-up (and 3D Tiles is z-up), the data must be accessed to reflect the transforms
       // from a Y-up to Z-up frame of reference, plus the Cesium -> Unreal transform as well.
-      LocalUV = float3(LocalUV.x, clamp(1.0 - LocalUV.z, 0.0, 1.0), LocalUV.y);
```
Comment (timoore) [inline Shaders/Private/CesiumVoxelTemplate.usf]: Seems to have snuck in from another PR...

File: Source/CesiumRuntime/Private/CesiumGltfComponent.cpp [hunk 3]
```diff
@@ -1438,8 +1437,7 @@ static void loadPrimitive(
   }
 
   const CesiumGltf::MaterialPBRMetallicRoughness& pbrMetallicRoughness =
-      material.pbrMetallicRoughness ? material.pbrMetallicRoughness.value()
-                                    : defaultPbrMetallicRoughness;
+      material.pbrMetallicRoughness.value_or(defaultPbrMetallicRoughness);
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfComponent.cpp:1440]: nice

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+
+  FPositionVertexBuffer& edgePositions =
+      lodResources.VertexBuffers.PositionVertexBuffer;
+  edgePositions.Init(totalEdgeVertices, false);
+
+  // Edge type is passed through as a UV coordinate in the edge vertex buffer.
+  // If silhouette edges are used, then two UV sets are used; the other three
+  // floats are used to pass the second silhouette normal for that edge. (See
+  // populateSilhouetteNormals)
+  int32 numTexCoords = 1 + int32(!visibleEdges.silhouetteEdgeIndices.empty());
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: My hacker heart loves treating booleans as 0 and 1, but I wonder if this would be clearer as
`int32 numTexCoords = visibleEdges.sillhouetteEdgeIndices.empty() ? 1 : 2;`? I had think about this for several minutes to get to the correct logic!

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+      outPositionB = FVector3f::Zero();
+      edgeVertexBuffer.SetVertexUV(edgeVertexIndexA, 0, FVector2f::Zero());
+      edgeVertexBuffer.SetVertexUV(edgeVertexIndexB, 0, FVector2f::Zero());
+      continue;
+    }
+
+    FVector3f positionA = positionView[a];
+    FVector3f positionB = positionView[b];
+
+    outPositionA.X = positionA.X * CesiumPrimitiveData::positionScaleFactor;
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: I suggest either a comment that you are doing both the Cesium-to-Unreal axis flip and the scaling, or a function in CesiumPrimitive.h that does that. Seems useful and would reduce repetition.

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+        FVector2f(float(edgeType)));
+    edgeVertexBuffer.SetVertexUV(
+        edgeVertexIndexB,
+        0,
+        FVector2f(float(edgeType)));
+  }
+
+  lodResources.IndexBuffer.SetIndices(
+      indices,
+      totalEdgeVertices >= std::numeric_limits<uint16>::max()
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: Why greater-than-or-equal-to and not greater-than?

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+  for (int64_t indexOffset = 0; indexOffset + 2 < indices.size();
+       indexOffset += 3) {
+    // Iterate over triangles (every three indices).
+    for (uint8_t vertex = 0; vertex < 3; vertex++) {
+      // Identify the next edge in the triangle.
+      // For each triangle (v0, v1, v2), the bitfield encodes three visibility
+      // values for the edges (v0:v1, v1:v2, v2:v0) in that order.
+      uint8_t nextVertex = (vertex + 1) % 3;
+      // Get the corresponding visibility value for the edge.
+      uint32 byteIndex = FMath::Floor(edgeIndex / 4);
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: What is `FMath::Floor` doing? This is integer division of a positive value (right?), and in any event doing a floor operation on an integer is a noop. Do you just mean to cast to `uint32`?

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+      uint32 bitPairOffset = (edgeIndex % 4) * 2;
```
Comment (timoore) [inline Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: `std::div` gives both quotient and remainder. It's probably not any faster, but I find it a bit cleaner than doing / and % in series.

**Review by timoore — COMMENTED** [review]
Looks good @j9liu ! Just some minor comments / questions.

File: Shaders/Private/CesiumVoxelTemplate.usf (hunk 2 above)
Comment (j9liu) [reply Shaders/Private/CesiumVoxelTemplate.usf]: Yeah, these changes are already in main, so I'm not sure why this is showing up in the diff 🤔

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp (hunk 7 above)
Comment (j9liu) [reply Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: Unthinkingly copied from CesiumJS, whoops :')

File: Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp (hunk 6 above)
Comment (j9liu) [reply Source/CesiumRuntime/Private/CesiumGltfPrimitiveEdges.cpp]: The max value for indices is technically disallowed by the glTF spec:

> indices accessor MUST NOT contain the maximum possible value for the component type used (i.e., 255 for unsigned bytes, 65535 for unsigned shorts, 4294967295 for unsigned ints).

Although the `EXT_primitive_restart` extension relaxes this, it's not clear whether Unreal supports this behavior, so we just fall back on the larger int type.

**Comment by j9liu** [comment]
Thanks @timoore! I believe I've addressed all your feedback. I also opened #1842 as a follow-up as well.

**Review by timoore — APPROVED** [review]
Looks fine to me now.
